lib/livebook/notebook/learn/github_stars.livemd

```diff
@@ -0,0 +1,331 @@
+
+# GitHub Stars
+
+```elixir
+Mix.install([
+  {:kino, "~> 0.15.3"},
+  {:req, "~> 0.5.10"},
+  {:kino_vega_lite, "~> 0.1.13"}
+])
+```
+
+## Introduction
+
+In this tutorial, we'll build a simple Livebook app together.
+
+We'll create an app with a form where we can enter a GitHub repository name, and it will show us how the repository's stars have grown over time - displayed both as a chart and as a table.
+
+Let's get started.
+
+## GitHub API client
+
+First, we'll build a simple API client for GitHub. We need this to fetch the list of people who have starred a repository.
+
+```elixir
+defmodule GitHubApi do
+  def stargazers(repo_name) do
+    with :ok <- rate_remaining(),
+         {:ok, response} <- do_stargazers(repo_name) do
+      maybe_paginate(response, repo_name)
+    end
+  end
+
+  defp rate_remaining do
+    response = Req.get!(base_req(), url: "/rate_limit")
+
+    if response.body["rate"]["remaining"] > 0 do
+      :ok
+    else
+      {:error, "GithubApi rate limit reached"}
+    end
+  end
+
+  defp base_req() do
+    Req.new(
+      base_url: "https://api.github.com",
+      headers: [
+        accept: "application/vnd.github.star+json",
+        "X-GitHub-Api-Version": "2022-11-28"
+      ]
+    )
+  end
+
+  defp do_stargazers(repo_name, page \\ 1) do
+    case Req.get(base_req(), url: "/repos/#{repo_name}/stargazers?per_page=100&page=#{page}") do
+      {:ok, %Req.Response{status: 200} = response} -> {:ok, response}
+      {:ok, %Req.Response{status: 403}} -> {:error, "GitHub API rate limit reached"}
+      {:ok, response} -> {:error, response.body["message"]}
+      {:error, exception} -> {:error, "Exception calling GitHub API: #{inspect(exception)}"}
+    end
+  end
+
+  defp maybe_paginate(response, repo_name) do
+    star_dates =
+      if "link" in Map.keys(response.headers) do
+        paginate(response, repo_name)
+      else
+        parse(response.body)
+      end
+
+    {:ok, star_dates}
+  end
+
+  defp paginate(response, repo_name) do
+    first_stargazers = parse(response.body)
+    last_page = last_page_number(response.headers)
+
+    additional_stargazers =
+      Task.async_stream(
+        2..last_page,
+        fn page -> do_stargazers(repo_name, page) end,
+        max_concurrency: 60
+      )
+      |> Enum.flat_map(fn
+        {:ok, {:ok, resp}} -> parse(resp.body)
+        _ -> []
+      end)
+
+    first_stargazers ++ additional_stargazers
+  end
+
+  defp parse(body) do
+    Enum.map(body, fn stargazer ->
+      %{"starred_at" => starred_at, "user" => %{"login" => user_login}} = stargazer
+      {:ok, starred_at, _} = DateTime.from_iso8601(starred_at)
+
+      %{
+        starred_at: starred_at,
+        user_login: user_login
+      }
+    end)
+  end
+
+  defp last_page_number(headers) do
+    links = hd(headers["link"])
+
+    %{"last_page" => last_page} =
+      Regex.named_captures(~r/<.*page=(?<last_page>\d+)>; rel="last"/, links)
+    String.to_integer(last_page)
+  end
+end
+```
```

Comment:
Unfortunately this module is too large, can we break it around perhaps? One to do request and another to paginate?

Reply:
Done in https://github.com/livebook-dev/livebook/pull/2990/commits/15c64c155bf88f44c249e61c09c69ae2bb267df4

Is it ok now?